packages/wallet/ui/src/util/keyManagement.js

```diff
@@ -0,0 +1,425 @@
+      action,
+      owner: toBech32(bech32Config.bech32PrefixAccAddr, fromBase64(owner)),
+    }),
+    fromAmino: ({ action, owner }) => ({
+      action,
+      owner: toBase64(toAccAddress(owner)),
+    }),
+  },
+  [SwingsetMsgs.MsgWalletSpendAction.typeUrl]: {
+    aminoType: SwingsetMsgs.MsgWalletAction.aminoType,
+    toAmino: ({ spendAction, owner }) => ({
+      spendAction,
+      owner: toBech32(bech32Config.bech32PrefixAccAddr, fromBase64(owner)),
+    }),
+    fromAmino: ({ spendAction, owner }) => ({
+      spendAction,
+      owner: toBase64(toAccAddress(owner)),
+    }),
+  },
+};
+
+/**
+ * key for use in localStorage
+ *
+ * arbitrary; suffix chosen randomly to avoid collisions
+ */
+export const STORAGE_KEY = 'agoric.eis0Aigi';
+
+/**
+ * Maintain a key for signing non-spending messages in localStorage.
+ *
+ * See also `delegateWalletAction()` below.
+ *
+ * @param {object} io
+ * @param {typeof window.localStorage} io.localStorage
+ * @param {typeof import('@cosmjs/crypto').Random.getBytes} io.getBytes for key generation
+ */
+export const makeBackgroundSigner = async ({ localStorage, getBytes }) => {
+  const provideLocalKey = () => {
+    const stored = localStorage.getItem(STORAGE_KEY);
+    if (stored) {
+      return fromBase64(stored);
+    }
+    console.log(
+      `localStorage.setItem(${STORAGE_KEY}, Random.getBytes(${KEY_SIZE}))`,
+    );
+    const seed = getBytes(KEY_SIZE);
+    localStorage.setItem(STORAGE_KEY, toBase64(seed));
+    return seed;
+  };
+  const seed = provideLocalKey();
+  const wallet = await DirectSecp256k1Wallet.fromKey(
+    seed,
+    bech32Config.bech32PrefixAccAddr,
+  );
+
+  const accounts = await wallet.getAccounts();
+  console.debug('device account(s):', accounts);
+
+  const [{ address }] = accounts;
+
+  /**
+   * Query grants for granter / grantee pair
+   *
+   * For example, to check whether `delegateWalletAction()` is necessary.
+   *
+   * @param {string} granter address
+   * @param {import('@cosmjs/tendermint-rpc').Tendermint34Client} rpcClient
+   */
+  const queryGrants = async (granter, rpcClient) => {
+    const base = QueryClient.withExtensions(rpcClient);
+    const rpc = createProtobufRpcClient(base);
+    const queryService = new QueryClientImpl(rpc);
+    console.log('query Grants', { granter, grantee: address });
+    const result = await queryService.Grants({
+      granter,
+      grantee: address,
+      msgTypeUrl: '', // wildcard
+    });
+    const decoded = result.grants.map(
+      g =>
+        g.authorization && GenericAuthorization.decode(g.authorization.value),
+    );
+
+    return decoded;
+  };
+
+  return freeze({
+    address,
+    registry: SwingsetRegistry,
+    wallet,
+    queryGrants,
+  });
+};
+
+/**
+ * @param {string} granter bech32 address
+ * @param {string} grantee bech32 address
+ * @param {number} seconds expiration as seconds (Date.now() / 1000)
+ */
+const makeGrantWalletActionMessage = (granter, grantee, seconds) => {
```

Comment:
I think CosmJS exports some message-specific facilities to make this kind of process simpler.  This is a fine approach until we sync better with the CosmJS team on our use cases.